Mathlib/Tactic/Basic.lean

```diff
@@ -31,6 +31,38 @@ syntax (name := lemma)
     stx.setKind ``Parser.Command.theorem
   pure <| stx.setKind ``Parser.Command.declaration
 
+/--
+Replace the type of the free variable `fvarId` with `typeNew`.
+
+If `checkDefEq = true` then throws an error if `typeNew` is not definitionally
+equal to the type of `fvarId`. Otherwise this function assumes `typeNew` and the type
+of `fvarId` are definitionally equal.
+
+This function is the same as `Lean.MVarId.changeLocalDecl` but makes sure to push substitution
+information into the infotree.
+-/
+def _root_.Lean.MVarId.changeLocalDecl' (mvarId : MVarId) (fvarId : FVarId) (typeNew : Expr)
+    (checkDefEq := true) : MetaM MVarId := do
+  mvarId.checkNotAssigned `changeLocalDecl
+  let (xs, mvarId) ← mvarId.revert #[fvarId] true
+  mvarId.withContext do
+    let check (typeOld : Expr) : MetaM Unit := do
+      if checkDefEq then
+        unless ← isDefEq typeNew typeOld do
+          throwTacticEx `changeLocalDecl mvarId
+            m!"given type{indentExpr typeNew}\nis not definitionally equal to{indentExpr typeOld}"
+    let finalize (targetNew : Expr) : MetaM MVarId := do
+      let mvarId ← mvarId.replaceTargetDefEq targetNew
+      let (ys, mvarId) ← mvarId.introNP xs.size
+      mvarId.withContext do
+        for x in xs, y in ys do
+          pushInfoLeaf (.ofFVarAliasInfo { id := y, baseId := x, userName := ← y.getUserName })
```

Comment:
I think it would be great to extract this snippet into a separate function (it already occurs twice in this PR)

Reply:
I made a little API for the revert/intro pattern that's hopefully useful beyond `change` and `clear_value`, and which is hopefully not over-engineered.

I extracted the snippet too, but it's not used in this API.